src/services/jsTyping.ts

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft. All rights reserved. Licensed under the Apache License, Version 2.0. 
+// See LICENSE.txt in the project root for complete license information.
+
+/// <reference path='services.ts' />
+
+/* @internal */
+namespace ts.JsTyping {
+
+    interface HostType {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+        writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void;
+    };
+
+    var _host: HostType;
+    var _safeList: Map<string>;
+
+    // a typing name to typing file path mapping
+    var inferredTypings: Map<string> = {};

```

Comment:
The `inferredTypings` is cleared every time a new inferring process is started (the `discoverTypings` is called), so the information it stores is not used either globally or by a project. Maybe I should clear the object at the end of the `discoverTypings` function to save memory, though that's not a structure issue.